### ISSUE TITLE
Still no auto-rollback feature

### ISSUE THREAD

**Issue by deliciouslytyped** [NixOS/nixpkgs#65477]
I did't find a separate issue for this yet, it's pretty much covered by the discussion in https://github.com/NixOS/nixpkgs/issues/34902 but may be a simpler starting point.

I would be happy if there was some official/upstreamed functionality for timed automatic rollback - e.g. if an ssh connection fails due to misconfiguration. This would mesh well with NixOS's friendlyness towards "safe changes" - i.e. you can "undo" things.

This would probably be covered by existing rollback scripts, I just want something officially endorsed.

**Comment by arianvp** [comment]
Systemd recently got support for automatic rollbacks when certain conditions aren't met. Perhaps we can build on top of that

**Comment by arianvp** [comment]
https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT

**Comment by deliciouslytyped** [comment]
I'd be fine for the initial implementation to just require the user to run a command to deactivate the reset.

**Comment by deliciouslytyped** [comment]
Boot related things seem more relevant to https://github.com/NixOS/nixpkgs/issues/26332 ? (Maybe needs a per-bootloader aggredation issue).

Or am I misunderstanding the intentions?

Or perhaps I was unclear? I meant for nixos-rebuild switch, not for reboots.

**Comment by aanderse** [comment]
The first step is to define criteria of when a `switch` is successful and when it is not. Probably not a difficult task, but needs some consideration.

**Comment by deliciouslytyped** [comment]
i'm imagining it with the user deciding if it was successful by turning off the reset timer instance.

**Comment by zimbatm** [comment]
The user could declare a healthcheck process that gets executed on boot. It could just be `sleep 300 && curl -sfL http://localhost:8080` for example. If the healthcheck fails the system would rollback.

**Comment by aanderse** [comment]
I was thinking you were intending something that could actually tell right away if the `switch` was successful or not. My mistake.

**Comment by lheckemann** [comment]
https://www.youtube.com/watch?v=J4DgATIjx9E talk by @basvandijk and @FPtje on how  LumiGuide does this.

**Comment by FPtje** [comment]
@lheckemann Yeah, I wrote that a while ago. One of the main benefits is that the condition of whether a switch was "good" is made by humans. A human has to tell the system that the upgrade was successful, otherwise it will automatically rollback after 20 minutes. This has saved us from being locked out of our systems quite a few times.

Some serious downsides to our system have prevented us from releasing it properly:
- Specific to our systems, requires some effort to migrate code to a more generic implementation
- It depends on nixos switch version numbers, which sometimes go funny ways when you're going back and forth between versions
- Rollback is done with `nixos-rebuild switch --rollback`. It would be more stable to switch using `nixos-rebuild test` and just reboot after 20 minutes, since switching configurations puts the device in a state "between" configurations. By that I mean services that aren't restarted, and a very specific Hetzner issue where switching would just kill the network under some circumstances. We had had that problem for quite a while, but it went away for us. I think @nh2 might still suffer from this problem.

**Comment by deliciouslytyped** [comment]
https://github.com/NixOS/nixpkgs/pull/84204 might be related.

**Comment by deliciouslytyped** [comment]
https://github.com/Infinisil/nixus looks like it has some sort of related functionality.

**Comment by FPtje** [comment]
That looks pretty clean actually, better than the systems services and timers stuff I had written 😅

**Comment by nh2** [comment]
still important

**Comment by hacklschorsch** [comment]
I recently almost screwed up an otherwise very nice dev machine by forcing the removal of a virtualbox QT web view something something dependency which trickled down all the way to the most basic tools like `ls` and nix itself.

I was lucky that I was already using the machine for a while and some multiple months old environments I could boot to were still lying around on the disk.

I would be much more comfortable if I wouldn't need luck to undo stupid mistakes like that.

**Comment by robintown** [comment]
Still a desirable feature

**Comment by Lucus16** [comment]
I'm going to see if I can tackle this in the coming weeks. I'm thinking a commit should consist of something like `cp /nix/var/nix/profiles/system /nix/var/nix/profiles/system-fallback` and a rollback should do the reverse, followed by `/nix/var/nix/profiles/system/bin/switch-to-configuration`. I'll probably make the choice of `switch` or `boot` a configuration choice as well.

**Comment by dasJ** [comment]
To me it looks like the easiest solution would be to solve this in `nixos-rebuild`:
- Read target of `/run/current-system` at the beginning
- Check that the action is not `switch` or `boot`
- After running stc, check if anything went wrong and use the store path from the first list item to go back

For remomte switches (not sure how well they are supported in `nixos-rebuild`) a nice bonus feature would be something like:
> systemd-run --description='switchback' --on-active=5m --service-type=oneshot ${oldSystem}/bin/switch-to-configuration ${action}

This would allow automatic rollback when something goes wrong AND the machine is not reachable anymore (would also work with reboot rather than rolling back live)
